Content.Server/GameObjects/Components/Observer/GhostComponent.cs

```diff
@@ -62,29 +76,82 @@ public override void HandleMessage(ComponentMessage message, IComponent componen
+                        if (Owner.TryGetComponent(out VisitingMindComponent mind))
+                        {
+                            Owner.EntityManager.EventBus.RaiseEvent(EventSource.Local, new GhostReturnMessage(mind.Mind));
+                        }
 
-                    break;
-                default:
-                    break;
-            }
+                        break;
+
+                    case GhostWarpRequestMessage warp:
+                        if (warp.PlayerTarget != default)
+                        {
+                            foreach (var player in _playerManager.GetAllPlayers())
+                            {
+                                if (warp.PlayerTarget == player.AttachedEntity.Uid)
+                                {
+                                    session.AttachedEntity.Transform.Coordinates =
+                                        player.AttachedEntity.Transform.Coordinates;
+                                }
+                            }
+                        }
+                        else
+                        {
+                            foreach (var warpPoint in FindWaypoints())
+                            {
+                                if (warp.WarpName == warpPoint.Location)
+                                {
+                                    session.AttachedEntity.Transform.Coordinates = warpPoint.Owner.Transform.Coordinates ;
+                                }
+                            }
+                        }
+                        break;
+
+                    case GhostRequestPlayerNameData data:
+                        var playerNames = new Dictionary<EntityUid, string>();
+                        foreach (var names in _playerManager.GetAllPlayers())
+                        {
+                            if (names.AttachedEntity != null && names.SessionId != netChannel.SessionId)
+                            {
+                                playerNames.Add(names.AttachedEntity.Uid,names.AttachedEntity.Name);
+                            }
+                        }
+                        ReplyGhostRequestPlayerNameData(playerNames);
+                        break;
+
+                    case GhostRequestWarpPointData data:
+
+                        var warpPoints = FindWaypoints();
+                        var warpName = warpPoints.Select(warpData => warpData.Location).ToList();
+                        ReplyGhostRequestWarpPoint(warpName);
+                        break;
+                    default:
+                        break;
+                }
+        }
+
+        private List<WarpPointComponent> FindWaypoints()
+        {
+            var comp = IoCManager.Resolve<IComponentManager>();
+            return comp.EntityQuery<WarpPointComponent>().ToList();
         }
 
+        private void ReplyGhostRequestPlayerNameData(Dictionary<EntityUid, string> playerNames) => SendNetworkMessage(new GhostReplyPlayerNameData(playerNames));
```

Comment:
idk if you need these as you only call them once

Reply:
What do you mean?

Reply:
Are those functions necessary? You could just move the `SendNetworkMessage` to where you call `ReplyGhostRequestPlayerNameData`.

Reply:
Either im not reading that sentence correctly or im just smooth brain _what_

Reply:
You do this:
```csharp
                        foreach (var names in _playerManager.GetAllPlayers())
                        {
                            if (names.AttachedEntity != null && names.SessionId != netChannel.SessionId)
                            {
                                playerNames.Add(names.AttachedEntity.Uid,names.AttachedEntity.Name);
                            }
                        }
                        ReplyGhostRequestPlayerNameData(playerNames);
                        break;
```
But the `ReplyGhostRequestPlayerNameData` isn't used anywhere else.
You could just
```csharp
                        foreach (var names in _playerManager.GetAllPlayers())
                        {
                            if (names.AttachedEntity != null && names.SessionId != netChannel.SessionId)
                            {
                                playerNames.Add(names.AttachedEntity.Uid,names.AttachedEntity.Name);
                            }
                        }
                        SendNetworkMessage(new GhostReplyPlayerNameData(playerNames))
                        break;
```
But you can also just leave this as is

Reply:
Ahh I see, ill get on it